src/store/mod.rs

```diff
@@ -0,0 +1,147 @@
+use crate::node_pages_map::{Page, PAGE_SIZE};
+use io_uring::{cqueue, opcode, squeue, types, IoUring};
+use slab::Slab;
+use std::{
+    fs::{File, OpenOptions},
+    io,
+    ops::Range,
+    os::{fd::AsRawFd, unix::fs::OpenOptionsExt},
+    path::PathBuf,
+    sync::Arc,
+};
+
+#[cfg(test)]
+mod tests;
+
+const RING_CAPACITY: u32 = 128;
+
+/// The Store is an on disk array of [`crate::node_pages_map::Page`]
+pub struct Store {
+    store_file: File,
+}
+
+/// Options to open a Store
+pub struct StoreOptions {
+    /// File path of the storage file.
+    file_path: PathBuf,
+}
+
+impl Default for StoreOptions {
+    fn default() -> Self {
+        Self {
+            file_path: PathBuf::from("node_pages_store"),
+        }
+    }
+}
+
+impl Store {
+    /// Create a new Store given the StoreOptions
+    pub fn new(options: StoreOptions) -> Self {
+        let store_file = OpenOptions::new()
+            .read(true)
+            .write(true)
+            .custom_flags(libc::O_DIRECT)
+            .open(options.file_path)
+            .expect("could not open file");
+
+        Store { store_file }
+    }
+}
+
+/// An I/O handle for a store. This owns an io_uring ring which is used to submit reads and writes.
+pub struct StoreIo {
+    store: Arc<Store>,
+    ring: IoUring,
+    pending: Slab<Pending>,
+}
+
+impl StoreIo {
+    pub fn new(store: Arc<Store>) -> Self {
+        StoreIo {
+            ring: IoUring::<squeue::Entry, cqueue::Entry>::builder()
+                .setup_sqpoll(100)
+                // .setup_iopoll
+                .build(RING_CAPACITY)
+                .expect("Error building io_uring"),
+            store,
+            pending: Slab::with_capacity(RING_CAPACITY as _),
```

Comment:
Please note one subtlety:

The number of items inflight is different from the ring capacity. Basically, when the kernel receives the request it removes it from the submission queue and, evidently, copies it to an in-kernel buffer.

Reply:
I have rewritten the code to use the slab as the main bottleneck as opposed to ring capacity.